### ISSUE TITLE
weird side-effect in tuple constructor

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#19352]
In an immutable wrapper around an `NTuple{8,UInt8}`, I'm getting a weird side-effect of `sort(collect(...))` that is mutating one of the tuple elements.  This happens both in 0.5 and master on MacOS X:
```jl
julia> immutable BitPerm
           p::NTuple{8,UInt8}
           function BitPerm(p::NTuple{8,UInt8})
               sort(collect(p)) != collect(0:7) && error("$p is not a permutation of 0:7")
               new(p)
           end
           BitPerm(b0,b1,b2,b3,b4,b5,b6,b7) = BitPerm((UInt8(b0),UInt8(b1),UInt8(b2),UInt8(b3),UInt8(b4),UInt8(b5),UInt8(b6),UInt8(b7)))
       end

julia> BitPerm(0,2,4,6,1,3,5,7)
BitPerm((0x00,0x90,0x04,0x06,0x01,0x03,0x05,0x07))
```
Notice that the second element is `0x90` rather than `0x02`.    If you comment out the `sort(collect(p)) != collect(0:7)` check, however, it gives `0x02` as expected.

This seems like it has to be a bug, since `sort(collect(p)) != collect(0:7)` should not have side effects, and in any case it should be impossible for a side effect to mutate a tuple.

**Comment by yuyichao** [comment]
Seems to be mac specific.

**Comment by yuyichao** [comment]
I can reproduce with this on an old rc3 build

```jl
@noinline g(a...) = ccall(:jl_breakpoint, Void, (Any,), a)

immutable BitPerm
    p::NTuple{8,UInt8}
    function BitPerm(p::NTuple{8,UInt8})
        g(g, g, p)
        new(p)
    end
end
@code_warntype BitPerm((0x0,0x2,0x4,0x6,0x1,0x3,0x5,0x7))
code_llvm(STDOUT, BitPerm, Tuple{NTuple{8,UInt8}}, false, true)
@code_native BitPerm((0x0,0x2,0x4,0x6,0x1,0x3,0x5,0x7))
@show BitPerm((0x0,0x2,0x4,0x6,0x1,0x3,0x5,0x7))
```

This seems to be a LLVM codegen bug. AFAICT, the relavant assembly is

```asm
        movq    (%r15), %rax
        movq    %rax, %r8
        shrq    $16, %r8
        movq    %rax, %r9
        shrq    $24, %r9
        movq    %rax, %rsi
        shrq    $32, %rsi
        movq    %rax, %rdi
        shrq    $40, %rdi
        movq    %rax, %rbx
        shrq    $48, %rbx
        movq    %rax, %rcx
        shrq    $56, %rcx
        movq    -80(%rbp), %rdx
        movq    %rdx, (%r14)
        movb    %al, (%r13)
        movb    %cl, 7(%r13)
        movb    %bl, 6(%r13)
        movb    %dil, 5(%r13)
        movb    %sil, 4(%r13)
        movb    %r9b, 3(%r13)
        movb    %r8b, 2(%r13)
        movb    %spl, 1(%r13)
        movq    %r13, %rax
```

where `r15` initially holds the source address and `r13` holds the sret address. LLVM somehow picks `spl` (lowest 8 bits of the stack pointer afaict) as the register to hold the problematic tuple element (which is problematic on it's own) but didn't emit any code to move the value into it (i.e. it should in principle have a `movq %rax, %rsp; shrq $8, %rsp` and somehow restore the stack pointer afterwards....).

Can you check if this is fixed in LLVM 3.9? If not, we should report this to LLVM.

**Comment by yuyichao** [comment]
From llc output it seems that `movb %spl, 1(%r13)` is meant to be `movb %ah, 1%(r13)`. Not sure why the JIT does so differently and why it doesn't happen with LLC.

**Comment by yuyichao** [comment]
Seems to be fixed in LLVM 3.9 for me...

**Comment by stevengj** [comment]
Great, thanks for checking.
